Issue/26506 Provides correct desciption in docstring that get_indexer methods are not yet supported

timgates42 opened the pull request:
- [x] fixes docstring for #26506 until implementation is ready
- [x] 46585 tests passed - none added as documentation change only
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
```
$ git diff upstream/master -u -- "*.py" | flake8 --diff && echo yes
yes
```
- [x] whatsnew entry
Provides correct desciption in docstring that Interval.get_indexer using any method other than default are not yet supported

timgates42 committed:
Started work on shared doc idea but noticed get_loc is already forked so will back out and follow that pattern

timgates42 committed:
Follow example of forking doc from get_loc method

Originally had made a template from the documentation so a common basis could be
used with the IntervalIndex adding a warning that the other methods were as yet
unsupported. In the end because I am not clear on what effect putting a template
in the core documentation dictionary I feel back to just following get_loc and
forking the doc string.

timgates42 committed:
Add back in note to adjust the doc string if fixing the implementation

timgates42 committed:
Add back in note to adjust the doc string if fixing the implementation

timgates42 commented:
I followed the example from get_loc to adjust the documentation, I originally had created a template the each documentation strings used as a basis but was concerned how it would interact with the global documentation dictionary and how it was used. Let me know if you would like any changes.

timgates42 commented:
Oh wow - find the right way to do it in the Contributors guide my apologies for rushing into it without checking the great documentation - I will fix the PR to follow the recomendation in https://pandas-docs.github.io/pandas-docs-travis/development/contributing_docstring.html#sharing-docstrings

timgates42 committed:
Use docstring substitution instead

Following the notes in the contributors guide - have opted for using docstring
substition
(https://pandas-docs.github.io/pandas-docs-travis/development/contributing_docstring.html#sharing-docstrings).

timgates42 committed:
Remove double full-stop

gfyoung commented:
cc @jreback

timgates42 committed:
Looks like this section has its own substition technique

timgates42 committed:
Drop unused import

WillAyd reviewed (changes requested):
I think it would be easier to just append a Raises section to the doctoring for IntervalIndex

WillAyd commented:
cc @jschendel

timgates42 commented:
Thanks - good suggestion will do

timgates42 committed:
Restore original

timgates42 committed:
Use Appender raises to notify of the deficiency

As suggested by @WillAyd instead of altering the docstring using a global
substitution just append a raises docstring section that warns about the parts
not yet implemented.

timgates42 commented:
Thanks @WillAyd it was much easier - have modified it to append a raises section to the docstring as requested.

timgates42 committed:
Resolve check_doc failure

Raises section must appear before Examples which complicates the use of just
appending a doc section and instead must perform a substitution.

pep8speaks commented:
Hello @timgates42! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/indexes/interval.py`](https://github.com/pandas-dev/pandas/blob/88f59c318648a30d1f45caf935c0d876f09630b1/pandas/core/indexes/interval.py):

> [Line 810:80](https://github.com/pandas-dev/pandas/blob/88f59c318648a30d1f45caf935c0d876f09630b1/pandas/core/indexes/interval.py#L810): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

timgates42 commented:
Ran into a complication with the plan as the Raises section must be before the Examples in the docstring so hopefully this Substitution technique is the appropriate answer.

timgates42 committed:
Satisfy flake8 overflow and line length

Avoid exceeding column count and maintain flake8 line overflow requirements

TomAugspurger commented on pandas/core/indexes/interval.py:
```diff
@@ -805,7 +807,15 @@ def get_value(self, series, key):
+                       **{'raises_section': textwrap.dedent("""\
+        Raises
+        ------
+        raises NotImplementedError if any method other than than the default
```
Typically this is formatted like

```
Raises
------
ExceptionName
    Condition.
```

timgates42 replied:
Thanks @TomAugspurger - have adjusted accordingly

timgates42 committed:
Use correct raises syntax - Thanks @TomAugspurger

timgates42 committed:
Substitution uses kwargs for dict formats.

If you supply an argument then it assumes the substitution is a straight
sequence for the string format as it uses args or kwargs.

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -648,6 +648,8 @@ def _check_method(self, method):
             return
 
         if method in ['bfill', 'backfill', 'pad', 'ffill', 'nearest']:
+            # The documentation warns these methods are not implemented so if
```
I don't think this comment actually adds anything here

timgates42 replied:
Ok I've dropped it.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2788,7 +2789,7 @@ def get_loc(self, key, method=None, tolerance=None):
             in the target are marked by -1.
 
-        Examples
+        %(raises_section)sExamples
```
should this not be on a separate line? (maybe remove the blank lines at the end of the template)

timgates42 replied:
Thanks @jreback that does look a lot cleaner - I've adjusted the substitution so its on its own line.

timgates42 committed:
Drop comment as requested by @jreback

timgates42 committed:
Cleaner look as suggested by @jreback

If we remove the line at the end when substituting in the raises section and add
one at the start then in the template the raises_section substitution can be on
its own line which is a cleaner solution - thanks to @jreback.

WillAyd commented:
@timgates42 I think this looks generally good. Can you post a screenshot of the rendered HTML doc?

timgates42 commented:
> @timgates42 I think this looks generally good. Can you post a screenshot of the rendered HTML doc?

Sure @WillAyd here it is

![image](https://user-images.githubusercontent.com/47873678/58418690-d3b3c080-80cb-11e9-973f-d19bb7fdd123.png)

WillAyd commented:
The Raises section looks off with the asterisks - do you know why that is appearing?

Sent from my iPhone

> On May 27, 2019, at 5:08 AM, Tim Gates <notifications@github.com> wrote:
> 
> @timgates42 I think this looks generally good. Can you post a screenshot of the rendered HTML doc?
> 
> Sure @WillAyd here it is
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

jreback commented:
ahh, just noticed @WillAyd comment, can you update.

timgates42 commented:
Good catch - according to the sphinx documentation this comes out if you are missing a colon (which I am) - I’ll add and regenerate - I actually copied this from another spot in pandas so presumably that needs fixing too - will find it

timgates42 commented:
Oh interesting I found the function I copied rename_categories in pandas/core/arrays/categorical.py and it the issue can be seen in the published doc http://pandas-docs.github.io/pandas-docs-travis/reference/api/pandas.CategoricalIndex.rename_categories.html?highlight=rename_categories#pandas.CategoricalIndex.rename_categories but the fix is not what I expected as pandas uses numpydoc a varient of sphinx doc and it looks like this raises syntax is common across pandas as is the strange asterisks sequence - http://pandas-docs.github.io/pandas-docs-travis/reference/api/pandas.api.types.union_categoricals.html?highlight=union_categoricals#pandas.api.types.union_categoricals

timgates42 commented:
I do get these warnings when producing the documentation which is the next lead to track down...
```
home/tgates/3rdparty/pandas/pandas/core/indexes/interval.py:docstring of pandas.IntervalIndex.get_indexer:54: WARNING: Inline strong start-string without end-string.
WARNING: toctree references unknown document 'reference/api/pandas.RangeIndex.start'
WARNING: toctree references unknown document 'reference/api/pandas.RangeIndex.stop'
WARNING: toctree references unknown document 'reference/api/pandas.RangeIndex.step'
WARNING: toctree references unknown document 'reference/api/pandas.Series.sparse'
```

WillAyd reviewed (approved):
@timgates42 thanks for the research. Given this is a pre-existing issue across the code base I'm OK with this PR as is and would take a follow up PR to fix the asterisks.

Of course up to @jreback as well

timgates42 commented:
Oh just spotted I had doubled up than and it was `any method other than than the default`, I'll fix it.

timgates42 committed:
Fix up use incorrect sentence that had a double "than" in it.

timgates42 committed:
Fix layout to match expected raises syntax

simonjayhawkins commented:
> Given this is a pre-existing issue across the code base I'm OK with this PR as is and would take a follow up PR to fix the asterisks.

without checking explicitly, it may be related to #26058, there seem to be a few issues regarding formatting. The dev docs at http://pandas-docs.github.io/pandas-docs-travis are being built on Travis using Sphinx v2.0.1

timgates42 commented:
Oh thanks for the heads up @simonjayhawkins I can confirm I built locally using sphinx v2.0.1 let me check 1.8.5 to see if it looks better.

timgates42 commented:
Rolling back to sphinx 1.8.5 didn't solve the issue, I'll continue investigation. I've setup a separate repo at https://github.com/timgates42/numpydoc-demo to run some testing of numpydoc to see if I can reproduce the issue. If I solve it then I'm happy to send another PR as its unrelated to the problem being addressed here. Let me know if anyone disagrees or wants a change to this PR otherwise assuming all is good please merge it if this is appropriate - Thanks.

jreback commented:
thanks @timgates42 

if you can create an issue for the astericks thing (you can just refer back to the comments here), and a PR to fix would be great!
